### DISCUSSION TITLE
LUA translate CC value to pitch-wheel value (14bit MSB LSB)

### DISCUSSION THREAD

**Discussion by damiensellier** [RomanKubiak/ctrlr#220] [Q&A] [answered]
Good evening,

I'm working on a small Ctrlr project to transform CC messages to Mackie MCU compatible messages.
It's a small pannel to allow midi devices to control DAWs via MCU protocol.

I send midi values with my controller from 0 to 127.
for the 8 precise moving faders, MCU needs "Pitch Wheel" values from the wide range -8192 to 8192 on channel 1 to 8 (fader 1 to 8)

if CC value is 0, the translated pitch wheel value is -8192
if CC value is 127 the translated pitch wheel value is  8192

This part is plain simple, the final decimal value is "ccValueOffset" _(ccValue * 128) -8192_

I struggle finding the right LUA formula to convert the given value,  to pitch-Wheel values with MSB LSB

This is the documentation : 

**Faders**

Received: Ei, ll, hh Move fader to position 
Transmitted: Ei, ll, hh Fader moved by user

i Fader ID (00 thru 07; Master = 08)
ll Fader position value low 7 bits (00–7F)
hh Fader position value high 7 bits (00–7F)

Example: E0, 40, 55 = Fader Ch. 1, position (55 << 7) + 40
Note: Message format for transmitted fader position is the same as for received position. 
Only the top (high) 10 of the 14 transmitted bits are required. 
Positions 0 to 1023 (decimal) are transmitted as 0000 to 03FF (Ei 00 00 to Ei 7F 7F).

**This is my script**

```
ccValueOffset = (ccValue*128) - 8192 -- Translate 1byte value from CC to 2bytes Pitch Wheel value like MCU
console("Displays the ccValueOffset : "..(ccValueOffset)) -- This one is OK

local bv=BigInteger(ccValueOffset)
local FaderValueLsb = bv:getBitRangeAsInt(0,7) -- position of lsb
local FaderValueMsb = bv:getBitRangeAsInt(7,7) -- position of msb

panel:sendMidiMessageNow(CtrlrMidiMessage({0xE0,FaderValueLsb, FaderValueMsb  })) -- Send Midi Message with  like MCU
```

Can you please help me with that, because The values  of the sent message are not good.

Thanks a lot for your help ;)

Damien

**Comment by damiensellier** [comment, marked as answer]
Edit:

I had it right at first but MidiMonitor shows me wrong values (-8192 to 8192) if the value is (0 to 16384).

this is the right one : 

```
ccValueOffset = ccValue*128  -- Translate 1byte value from CC to 2bytes Pitch Wheel value like MCU
console("Displays the ccValueOffset : "..(ccValueOffset)) -- This one is OK

local bv=BigInteger(ccValueOffset)
local FaderValueLsb = bv:getBitRangeAsInt(0,7) -- position of lsb
local FaderValueMsb = bv:getBitRangeAsInt(7,7) -- position of msb

panel:sendMidiMessageNow(CtrlrMidiMessage({0xE0,FaderValueLsb, FaderValueMsb  })) -- Send Midi Message like MCU
```

Now I just need to change the last nibble of the first byte with the correct midi channel related to the fader #i : 0xEi

**Comment by dobo365** [comment]
Salut Damien!

Your basic equation is wrong. With
_if CC value is 0, the translated pitch wheel value is -8192
if CC value is 127 the translated pitch wheel value is 8192_

The ccValueOffset calculation is  ccValueOffset = (ccValue*16384/127) - 8192

I would also not use BigInteger and co...

I struggled a bit with the conversion (nice morning exercice - thank you) but I think that this should work:
```
ccValue=0 -- this is your input value
console("ccValue: "..tostring(ccValue))
ccValueOffset = (ccValue*16384/127) - 8192 -- Translate 1byte value from CC to 2bytes Pitch Wheel value like MCU
console("ccValueOffset : "..tostring(ccValueOffset))
ccMSB = bit.band(bit.rshift(ccValueOffset, 8), 0xF0)
ccLSB = bit.band(ccValueOffset, 0x7F)
console(string.format("%.2x", ccMSB))
console(string.format("%.2x", ccLSB))
```

**Reply by damiensellier** [reply]
Salut Goodweather!
I tried with the bit.band first but never had the right value as well.

As said in my reply before, I made a mistake, There's no need to remove 8192
I based my calculation on those damn values in midiMonitor, since it's a pitch wheel and has Left center right actions, the value was written negative for pitching left and positive for right which is BAD :) and made me lose my mind for 3 hours.

I tried and everything is working really well now, this is what I have : 

```
	ccValueOffset = (ccValue*128) -- Translate 1bit Midi Message with CC value to 2bits Pitch Wheel value like MCU

	local bv=BigInteger(ccValueOffset)
	local FaderValueLsb = bv:getBitRangeAsInt(0,7) -- position of lsb
	local FaderValueMsb = bv:getBitRangeAsInt(7,7) -- position of msb

	if ccValueOffset == 16256 then   -- Set LSB to 127 to reach Max Value 16256 instead of 16128
		FaderValueLsb = 127
	end

	panel:sendMidiMessageNow(CtrlrMidiMessage({0xE0,FaderValueLsb, FaderValueMsb })) -- Send Midi Message
```

Hope everything is alright on your side, I was very busy this winter so I start working back on Ctrlr pannels slowly but surely :)
